crates/sui-core/src/authority_active.rs

```diff
@@ -176,33 +181,185 @@ impl<A> ActiveAuthority<A>
+        gossip: bool,
+        checkpoint: bool,
+    ) {
+        let gossip_receiver = receiver.clone();
         let gossip_locals = self.clone();
-        let _gossip_join = tokio::task::spawn(async move {
-            if gossip {
-                gossip_process(&gossip_locals, 4).await;
+        let gossip_join = gossip_receiver.spawn("Gossip".to_string(), move || {
+            let inner_locals = gossip_locals.clone();
+            async move {
+                if gossip {
+                    gossip_process(&inner_locals, 4).await;
+                }
+                Ok(())
             }
         });
 
-        // Spawn task to take care of checkpointing
         let checkpoint_locals = self; // .clone();
-        let _checkpoint_join = tokio::task::spawn(async move {
-            if checkpoint {
-                checkpoint_process(&checkpoint_locals, &CheckpointProcessControl::default()).await;
+        let checkpoint_join = receiver.spawn("Gossip".to_string(), move || {
+            let inner_locals = checkpoint_locals.clone();
+            async move {
+                if checkpoint {
+                    checkpoint_process(&inner_locals, &CheckpointProcessControl::default()).await;
+                }
+                Ok(())
             }
         });
 
-        if let Err(err) = _gossip_join.await {
-            error!("Join gossip task end error: {:?}", err);
-        }
-        if let Err(err) = _checkpoint_join.await {
-            error!("Join checkpoint task end error: {:?}", err);
+        join!(gossip_join, checkpoint_join);
+    }
+}
+
+/*
+    The active authority start-update-shutdown facilities.
+
+    The active authority runs in a separate task, and manages the tasks of separate active
+    processes, with one sub-task per active process. These subtasks internally listen to a
+    watch channel on which others may send a start, shutdown, or update signal. When this
+    is received the tasks are restarted, or shut down.
+
+*/
+
+use tokio::sync::watch;
+
+pub enum LifecycleSignal {
+    Start,
+    Restart,
+    Exit,
+}
+
+pub struct LifecycleSignalSender {
+    sender: watch::Sender<LifecycleSignal>,
+}
+
+impl LifecycleSignalSender {
+    pub fn new() -> (Self, LifecycleTaskHandler) {
+        let (sender, receiver) = watch::channel(LifecycleSignal::Start);
+        (
+            LifecycleSignalSender { sender },
+            LifecycleTaskHandler { receiver },
+        )
+    }
+
+    pub async fn signal(&self, signal: LifecycleSignal) {
+        // TODO: reflect on what to do with this error.
+        let _ = self.sender.send(signal);
+    }
+}
+
+#[derive(Clone)]
+pub struct LifecycleTaskHandler {
+    pub receiver: watch::Receiver<LifecycleSignal>,
+}
+
+impl LifecycleTaskHandler {
+    pub async fn spawn<F, T>(mut self, description: String, fun: F)
+    where
+        F: Fn() -> T,
+        T: Future + Send + 'static,
+        <T as Future>::Output: Borrow<Result<(), SuiError>> + Send + 'static,
```

Comment:
I don't think you should need the `Borrow` wrapper around this

Reply:
So how do I tell the bound that Output should be `Result<(), SuiError>`?

Reply:
```
T: Future<Output = Result<(), SuiError>> + Send + 'static,
```

Reply:
Ok. I worked it out :)